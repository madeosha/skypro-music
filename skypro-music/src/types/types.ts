export type SignInFormType = {
  email: string;
  password: string;
};

export type SignUpFormType = {
  email: string;
  password: string;
  username: string;
};

export type UserType = {
  id: number;
  username: string;
  firstName: string;
  lastName: string;
  email: string;
};
