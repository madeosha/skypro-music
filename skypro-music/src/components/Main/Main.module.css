.wrapper {
    width: 100%;
    min-height: 100%;
    overflow: hidden;
    background-color: #383838;
}

.container {
    max-width: 1920px;
    height: 100vh;
    margin: 0 auto;
    position: relative;
    background-color: #181818;
}

.main {
    -webkit-box-flex: 1;
    -ms-flex: 1 1 auto;
    flex: 1 1 auto;
    display: -webkit-box;
    display: -ms-flexbox;
    display: flex;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    -webkit-box-pack: justify;
    -ms-flex-pack: justify;
    justify-content: space-between;
}

._btn_text:hover {
    border-color: #d9b6ff;
    color: #d9b6ff;
    cursor: pointer;
}

._btn_icon:hover svg {
    fill: transparent;
    stroke: #acacac;
    cursor: pointer;
}

._btn_text:active {
    border-color: #ad61ff;
    color: #ad61ff;
    cursor: pointer;
}

._btn_icon:active svg {
    fill: transparent;
    stroke: #ffffff;
    cursor: pointer;
}

._btn_icon:active .track_play__like_svg,
._btn_icon:active .track_play__dislike_svg {
    fill: #696969;
    stroke: #ffffff;
    cursor: pointer;
}