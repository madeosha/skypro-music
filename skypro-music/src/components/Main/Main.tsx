//"use client";
//import React, { useEffect, useMemo, useState } from "react";
//import { Menu } from "../Menu/Menu";
//import { Sidebar } from "../Sidebar/Sidebar";
//import styles from "./Main.module.css";
//import { Track } from "./Main.types";
//import { useAppSelector } from "@/store/store";
//import { Player } from "../Player/Player";
//import { useDispatch } from "react-redux";
//import { setCurrentTracks, setTracks } from "../../store/features/playerSlice";
//import Centerblock from "../Centerblock/Centerblock";

//type Main = {
//  tracks: Track[];
//};

//export const Main = ({ tracks }: Main) => {
//  const dispatch = useDispatch();

//  useEffect(() => {
//    dispatch(setTracks(tracks));
//    dispatch(setCurrentTracks(tracks));
//  }, [tracks, dispatch]);

//  return (
//    <div className={styles.wrapper}>
//      <div className={styles.container}>
//        <main className={styles.main}>
//          <Menu />
//          <Centerblock />
//          <Sidebar />
//        </main>
//        {currentTrack && <Player />}
//        <footer className={styles.footer}></footer>
//      </div>
//    </div>
//  );
//};
