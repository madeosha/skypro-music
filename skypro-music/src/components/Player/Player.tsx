"use client";
import classNames from "classnames";
import styles from "./Player.module.css";
import React, { useEffect, useRef, useState } from "react";
import { ProgressBar } from "../ProgressBar/ProgressBar";
import { useAppSelector } from "../../store/store";
import { useDispatch } from "react-redux";
import {
  setCurrentTrack,
  setIsPlaying,
  setIsShuffle,
} from "../../store/features/playerSlice";
import { useLikeTrack } from "../../hooks/likeTrack";

export const Player = () => {
  // Получаем ссылку на DOM-элемент audio
  const audioRef = useRef<HTMLAudioElement>(null);
  const audio = audioRef.current;
  const dispatch = useDispatch();
  // Вытаскивает текущий плейлист из глобального состояния
  const currentTrackList = useAppSelector(
    (state) => state.player.currentPlaylist
  );
  // Вытаскивает оригинальный плейлист из глобального состояния
  const originTrackList = useAppSelector((state) => state.player.currentTrack);
  // Вытаскиваем состояние текущего трека
  const currentTrack = useAppSelector((state) => state.player.currentTrack);
  const { isLiked, handleLike } = currentTrack
    ? useLikeTrack(currentTrack)
    : { isLiked: false, handleLike: () => {} };
  // Находим индекс текущего трека
  const currentTrackIndex = currentTrackList.findIndex(
    (track) => track.id === currentTrack?.id
  );
  // Состояние плеера из глобального состояния
  const isPlaying = useAppSelector((state) => state.player.isPlaying);
  // Состояние режима перемешивания из глобального состояния
  const isShuffle = useAppSelector((state) => state.player.isShuffle);
  // Состояние громкости
  const [isVolume, setIsVolume] = useState("0.5");
  // Состояние текущего времени
  const [currentTime, setCurrentTime] = useState(0);
  // Состояние повтора
  const [isLoop, setIsLoop] = useState(false);
  const duration: number = audio?.duration || 0;
  audio ? (audio.loop = isLoop) : null;
  // Функция для переключения следующего трека
  const handleEnded = () => {
    nextTrackClick();
  };
  // Функция зацикливания трека
  const toggleLoop = () => {
    setIsLoop(!isLoop);
  };
  // Функция для включения режима перемешивания
  const toggleShuffle = () => {
    dispatch(setIsShuffle(!isShuffle));
  };
  // Функция для воспроизведения и паузы
  const togglePlay = () => {
    if (audio) {
      if (isPlaying) {
        audio.pause();
      } else {
        audio.play();
      }
    }
    dispatch(setIsPlaying(!isPlaying));
  };
  // Функция для воспроизведения следующего трека
  const nextTrackClick = () => {
    if (currentTrackIndex < currentTrackList.length - 1) {
      const nextTrack = currentTrackList[currentTrackIndex + 1];
      dispatch(setCurrentTrack(nextTrack));
    }
  };
  // Функция для воспроизведения предыдущего трека
  const prevTrackClick = () => {
    if (currentTrackIndex !== 0) {
      const prevTrack = currentTrackList[currentTrackIndex - 1];
      dispatch(setCurrentTrack(prevTrack));
    }
  };
  // Меняем громкость при изменении ползунка громкости
  useEffect(() => {
    if (audio) {
      audio.volume = parseFloat(isVolume);
    }
  }, [isVolume, audio]); // была ошибка
  useEffect(() => {
    const audio = audioRef.current;
    audio?.addEventListener("ended", handleEnded);
    dispatch(setIsPlaying(true));
    audio?.play();
    return () => audio?.removeEventListener("ended", handleEnded);
  }, [currentTrack]);

  return (
    <div className={styles.bar}>
      <audio
        ref={audioRef}
        src={currentTrack?.track_file}
        onTimeUpdate={(e) => setCurrentTime(e.currentTarget.currentTime)}
      />
      <div className={styles.bar__content}>
        <ProgressBar
          currentTime={currentTime}
          duration={duration}
          audioRef={audioRef.current}
        />
        <div className={styles.bar__player_block}>
          <div className={styles.bar__player}>
            <div className={styles.player__controls}>
              <div className={styles.player__btn_prev}>
                <svg
                  onClick={prevTrackClick}
                  className={classNames(
                    styles.player__btn_prev_svg,
                    styles._btn
                  )}
                >
                  <use href="/img/icon/sprite.svg#icon-prev"></use>
                </svg>
              </div>
              <div className={styles.player__btn_play}>
                <svg
                  onClick={togglePlay}
                  className={classNames(
                    styles.player__btn_play_svg,
                    styles._btn
                  )}
                >
                  {isPlaying ? (
                    <use href="/img/icon/sprite.svg#icon-pause"></use>
                  ) : (
                    <use href="/img/icon/sprite.svg#icon-play"></use>
                  )}
                </svg>
              </div>
              <div className={styles.player__btn_next}>
                <svg
                  data-testid="next-track-button"
                  onClick={nextTrackClick}
                  className={classNames(
                    styles.player__btn_next_svg,
                    styles._btn
                  )}
                >
                  <use href="/img/icon/sprite.svg#icon-next"></use>
                </svg>
              </div>
              <div
                className={classNames(
                  styles.player__btn_repeat,
                  styles._btn_icon
                )}
              >
                <svg
                  onClick={toggleLoop}
                  className={classNames(
                    styles.player__btn_repeat_svg,
                    styles._btn,
                    isLoop ? styles.player__btn_repeat_svg_active : null
                  )}
                >
                  <use href="/img/icon/sprite.svg#icon-repeat"></use>
                </svg>
              </div>
              <div
                className={classNames(
                  styles.player__btn_shuffle,
                  styles._btn_icon
                )}
              >
                <svg
                  onClick={toggleShuffle}
                  className={classNames(
                    styles.player__btn_shuffle_svg,
                    isShuffle ? styles.player__btn_shuffle_svg_active : null
                  )}
                >
                  <use href="/img/icon/sprite.svg#icon-shuffle"></use>
                </svg>
              </div>
            </div>

            <div
              className={classNames(
                styles.player__track_play,
                styles.track_play
              )}
            >
              <div className={styles.track_play__contain}>
                <div className={styles.track_play__image}>
                  <svg className={styles.track_play__svg}>
                    <use href="/img/icon/sprite.svg#icon-note"></use>
                  </svg>
                </div>
                <div className={styles.track_play__author}>
                  <a className={styles.track_play__author_link} href="http://">
                    {currentTrack?.name}
                  </a>
                </div>
                <div className={styles.track_play__album}>
                  <a className={styles.track_play__album_link} href="http://">
                    {currentTrack?.author}
                  </a>
                </div>
              </div>

              <div className={styles.track_play__like_dis}>
                <div
                  className={classNames(
                    styles.track_play__like,
                    styles._btn_icon
                  )}
                >
                  <svg
                    className={styles.track_play__like_svg}
                    onClick={handleLike}
                  >
                    <use
                      href={
                        isLiked
                          ? "/img/icon/sprite.svg#icon-dislike"
                          : "/img/icon/sprite.svg#icon-like"
                      }
                    ></use>
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div className={classNames(styles.bar__volume_block, styles.volume)}>
            <div className={styles.volume__content}>
              <div className={styles.volume__image}>
                <svg className={styles.volume__svg}>
                  <use href="/img/icon/sprite.svg#icon-volume"></use>
                </svg>
              </div>
              <div className={classNames(styles.volume__progress, styles._btn)}>
                <input
                  className={classNames(
                    styles.volume__progress_line,
                    styles._btn
                  )}
                  type="range"
                  min="0"
                  max="1"
                  step="0.01"
                  value={isVolume}
                  onChange={(e) => setIsVolume(e.target.value)}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};
