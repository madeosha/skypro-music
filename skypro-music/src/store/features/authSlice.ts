import { fetchAddUser, fetchTokens, fetchUser } from "../../api/userApi";
import { SignInFormType, SignUpFormType, UserType } from "../../types/types";
import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";

export const getAuthUser = createAsyncThunk(
  "user/getAuthUser",
  async ({ email, password }: SignInFormType) => {
    const user = await fetchUser({ email, password });
    return user;
  }
);

export const getRegUser = createAsyncThunk(
  "user/getRegUser",
  async ({ email, password, username }: SignUpFormType) => {
    const user = await fetchAddUser({ email, password, username });
    return user;
  }
);

export const getTokens = createAsyncThunk(
  "user/getTokens",
  async ({ email, password }: SignInFormType) => {
    const tokens = await fetchTokens({ email, password });
    return tokens;
  }
);

type AuthStateType = {
  user: null | UserType;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
};

const initialState: AuthStateType = {
  user: null,
  tokens: {
    access: null,
    refresh: null,
  },
};

const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    setLogout: (state) => {
      state.user = null;
      state.tokens.access = null;
      state.tokens.refresh = null;
    },
  },
  extraReducers(builder) {
    builder
      .addCase(
        getAuthUser.fulfilled,
        (state, action: PayloadAction<UserType>) => {
          state.user = action.payload;
        }
      )
      .addCase(
        getTokens.fulfilled,
        (
          state,
          action: PayloadAction<{
            access: string | null;
            refresh: string | null;
          }>
        ) => {
          state.tokens.access = action.payload.access;
          state.tokens.refresh = action.payload.refresh;
        }
      );
  },
});

export const { setLogout } = authSlice.actions;
export const authReducer = authSlice.reducer;
