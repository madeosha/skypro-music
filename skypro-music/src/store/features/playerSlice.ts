import { Track } from "../../components/Main/Main.types";
import { createSlice, PayloadAction } from "@reduxjs/toolkit";

type playerStateType = {
  isPlaying: boolean;
  currentTrack: Track | null;
  playlist: Track[];
  isShuffle: boolean;
  filterOptions: {
    author: string[];
    genre: string[];
    order: string;
    searchString: string;
  };
  filterPlaylist: Track[];
  currentPlaylist: Track[];
};

const initialState: playerStateType = {
  isPlaying: false,
  currentTrack: null,
  playlist: [],
  isShuffle: false,
  filterOptions: {
    author: [],
    genre: [],
    order: "По умолчанию",
    searchString: "",
  },
  filterPlaylist: [],
  currentPlaylist: [],
};

const playerSlice = createSlice({
  name: "player",
  initialState,
  reducers: {
    setTracks: (state, action: PayloadAction<Track[]>) => {
      state.playlist = action.payload;
      state.filterPlaylist = action.payload;
    },
    setCurrentTracks: (state, action: PayloadAction<Track[]>) => {
      state.currentPlaylist = action.payload;
    },
    setCurrentTrack: (state, action: PayloadAction<Track>) => {
      state.currentTrack = action.payload;
    },
    setIsPlaying: (state, action: PayloadAction<boolean>) => {
      state.isPlaying = action.payload;
    },
    setIsShuffle: (state, action: PayloadAction<boolean>) => {
      state.isShuffle = action.payload;
      if (state.isShuffle) {
        const copyFilterPlaylist = [...state.filterPlaylist];
        state.currentPlaylist = copyFilterPlaylist.sort(
          () => 0.5 - Math.random()
        );
      } else {
        state.currentPlaylist = state.filterPlaylist;
      }
    },
    setFilters: (
      state,
      action: PayloadAction<{
        author?: string[];
        genre?: string[];
        order?: string;
        searchString?: string;
      }>
    ) => {
      state.filterOptions = {
        author: action.payload.author || state.filterOptions.author,
        genre: action.payload.genre || state.filterOptions.genre,
        order: action.payload.order || state.filterOptions.order,
        searchString:
          action.payload.searchString || state.filterOptions.searchString,
      };

      const filterTracks = [...state.playlist].filter((track) => {
        const hasSearchString = track.name
          .toLowerCase()
          .includes(state.filterOptions.searchString.toLowerCase());
        // Если выбрали фильтр по автору, то проверяем трек на соотвествие этим авторам
        // Если фильтр по автору не выбран, то возвращаем этот трек, так как нам нужно вывести все треки
        const hasAuthor =
          state.filterOptions.author.length > 0
            ? state.filterOptions.author.includes(track.author)
            : true;
        const hasGenre =
          state.filterOptions.genre.length > 0
            ? state.filterOptions.genre.includes(track.genre)
            : true;
        return hasSearchString && hasAuthor && hasGenre;
      });

      switch (state.filterOptions.order) {
        case "Сначала новые":
          filterTracks.sort(
            (a, b) =>
              new Date(b.release_date).getTime() -
              new Date(a.release_date).getTime()
          );
          break;
        case "Сначала старые":
          filterTracks.sort(
            (a, b) =>
              new Date(a.release_date).getTime() -
              new Date(b.release_date).getTime()
          );

        default:
          break;
      }
      state.filterPlaylist = filterTracks;
      state.currentPlaylist = filterTracks;
    },
  },
});

export const {
  setTracks,
  setCurrentTracks,
  setCurrentTrack,
  setIsPlaying,
  setIsShuffle,
  setFilters,
} = playerSlice.actions;

export const playerReducer = playerSlice.reducer;
