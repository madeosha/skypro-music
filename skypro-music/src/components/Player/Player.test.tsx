import React, { act } from "react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import { Player } from "./Player";
import { Track } from "../Main/Main.types";
import { setCurrentTrack, setIsPlaying } from "@/store/features/playerSlice";

// Тип для корневого состояния Redux
interface RootState {
  player: {
    currentPlaylist: Track[];
    currentTrack: Track | null;
    isPlaying: boolean;
    isShuffle: boolean;
  };
}

const mockStore = configureStore<RootState>([]);

describe("Корректный рендер компонента Player", () => {
  let store: MockStoreEnhanced<RootState>;
  let initialState: RootState;

  beforeEach(() => {
    initialState = {
      player: {
        currentPlaylist: [
          {
            id: 1,
            name: "Track 1",
            author: "Artist 1",
            release_date: "2022-01-01",
            genre: "Genre 1",
            duration_in_seconds: 210,
            album: "Album 1",
            logo: "/path/to/logo1.png",
            track_file: "/path/to/track1.mp3",
            stared_user: [],
          },
          {
            id: 2,
            name: "Track 2",
            author: "Artist 2",
            release_date: "2022-01-02",
            genre: "Genre 2",
            duration_in_seconds: 180,
            album: "Album 2",
            logo: "/path/to/logo2.png",
            track_file: "/path/to/track2.mp3",
            stared_user: [],
          },
        ],
        currentTrack: {
          id: 1,
          name: "Track 1",
          author: "Artist 1",
          release_date: "2022-01-01",
          genre: "Genre 1",
          duration_in_seconds: 210,
          album: "Album 1",
          logo: "/path/to/logo1.png",
          track_file: "/path/to/track1.mp3",
          stared_user: [],
        },
        isPlaying: false,
        isShuffle: false,
      },
    };

    store = mockStore(initialState);
  });

  it("Корректный рендер компонента Player", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Player />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Переключение следующего трека", () => {
    store.dispatch(setIsPlaying(true));

    const component = renderer.create(
      <Provider store={store}>
        <Player />
      </Provider>
    );

    // Имитируем клик на кнопку следующего трека
    const nextTrackClick = component.root.findByProps({
      "data-testid": "next-track-button",
    });
    act(() => {
      nextTrackClick.props.onClick();
    });
    // Ожидаем, что после клика будет вызван setCurrentTrack с ожидаемым следующим треком
    const expectedNextTrack = store.getState().player.currentPlaylist[1];
    const actions = store.getActions();
    const setCurrentTrackAction = actions.find(
      (action) => action.type === setCurrentTrack.type
    );
    expect(setCurrentTrackAction.payload).toEqual(expectedNextTrack);
  });
});
