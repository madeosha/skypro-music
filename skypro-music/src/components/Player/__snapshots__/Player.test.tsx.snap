// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Корректный рендер компонента Player Корректный рендер компонента Player 1`] = `
<div
  className="bar"
>
  <audio
    onTimeUpdate={[Function]}
    src="/path/to/track1.mp3"
  />
  <div
    className="bar__content"
  >
    <div
      className="bar__player_progress"
    >
      <div
        className="bar__currentTime"
      >
        0:00
         / 
        0:00
      </div>
      <input
        className="styledProgressInput"
        max={0}
        min={0}
        onChange={[Function]}
        step={0.01}
        type="range"
        value={0}
      />
    </div>
    <div
      className="bar__player_block"
    >
      <div
        className="bar__player"
      >
        <div
          className="player__controls"
        >
          <div
            className="player__btn_prev"
          >
            <svg
              className="player__btn_prev_svg _btn"
              onClick={[Function]}
            >
              <use
                href="/img/icon/sprite.svg#icon-prev"
              />
            </svg>
          </div>
          <div
            className="player__btn_play"
          >
            <svg
              className="player__btn_play_svg _btn"
              onClick={[Function]}
            >
              <use
                href="/img/icon/sprite.svg#icon-play"
              />
            </svg>
          </div>
          <div
            className="player__btn_next"
          >
            <svg
              className="player__btn_next_svg _btn"
              data-testid="next-track-button"
              onClick={[Function]}
            >
              <use
                href="/img/icon/sprite.svg#icon-next"
              />
            </svg>
          </div>
          <div
            className="player__btn_repeat _btn_icon"
          >
            <svg
              className="player__btn_repeat_svg _btn"
              onClick={[Function]}
            >
              <use
                href="/img/icon/sprite.svg#icon-repeat"
              />
            </svg>
          </div>
          <div
            className="player__btn_shuffle _btn_icon"
          >
            <svg
              className="player__btn_shuffle_svg"
              onClick={[Function]}
            >
              <use
                href="/img/icon/sprite.svg#icon-shuffle"
              />
            </svg>
          </div>
        </div>
        <div
          className="player__track_play track_play"
        >
          <div
            className="track_play__contain"
          >
            <div
              className="track_play__image"
            >
              <svg
                className="track_play__svg"
              >
                <use
                  href="/img/icon/sprite.svg#icon-note"
                />
              </svg>
            </div>
            <div
              className="track_play__author"
            >
              <a
                className="track_play__author_link"
                href="http://"
              >
                Track 1
              </a>
            </div>
            <div
              className="track_play__album"
            >
              <a
                className="track_play__album_link"
                href="http://"
              >
                Artist 1
              </a>
            </div>
          </div>
          <div
            className="track_play__like_dis"
          >
            <div
              className="track_play__like _btn_icon"
            >
              <svg
                className="track_play__like_svg"
              >
                <use
                  href="/img/icon/sprite.svg#icon-like"
                />
              </svg>
            </div>
            <div
              className="track_play__dislike _btn_icon"
            >
              <svg
                className="track_play__dislike_svg"
              >
                <use
                  href="/img/icon/sprite.svg#icon-dislike"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="bar__volume_block volume"
      >
        <div
          className="volume__content"
        >
          <div
            className="volume__image"
          >
            <svg
              className="volume__svg"
            >
              <use
                href="/img/icon/sprite.svg#icon-volume"
              />
            </svg>
          </div>
          <div
            className="volume__progress _btn"
          >
            <input
              className="volume__progress_line _btn"
              max="1"
              min="0"
              onChange={[Function]}
              step="0.01"
              type="range"
              value="0.5"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
